lib/dictBuilder/zdict.c

```diff
@@ -48,6 +48,7 @@
 #  define ZDICT_STATIC_LINKING_ONLY
 #endif
 #include "zdict.h"
+#include "decompress/zstd_decompress_internal.h" /* ZSTD_entropyDTables_t */
```

Comment:
In https://github.com/facebook/zstd/tree/dev/lib#modular-build, we have specified the following dependency graph : 

> lib/dictBuilder : (...) This module depends on both `lib/common` and `lib/compress` .

This `#include` changes that, by making `zdict` now _also_ dependent on `lib/decompress`.

I believe this is significant. It would for example break the build of `dictBuilder` if `lib/decompress` is not present.
(btw, it also shows that we are probably missing one test about this possible build).

I would prefer the dependency graph to not change.

Potential work-around (for discussion) : 
- Move `ZSTD_entropyDTables_t` definition somewhere else, likely in `lib/common`, for example `zstd_internal.h`. Make sure to add code comments to explain why it's there.
- More complex, and debatable : move `ZDICT_getDictHeaderSize` into `lib/common`, though it would also change its name (`ZSTD_*`) and its exposition path (through `zstd.h`). It also still requires `ZSTD_entropyDTables_t` in `lib/common`, since `lib/common` _must not_ depend on `lib/decompress`.

Reply:
Ah, I did wonder if there were some specification limitations with what `zdict.c` could `#include`. I think the first option is better - I'd feel better with this in the `ZDICT_*` API since it's use-case isn't as generalized as those in `ZSTD_*`, and it is a more expensive function than it would seem at first glance.

However, another issue is that `ZSTD_loadDEntropy()` also would need to be moved to `zstd_internal.h`. So in total, we will be moving the symbols `ZSTD_loadDEntropy()`, `ZSTD_seqSymbol`, `ZSTD_entropyDTables_t`, and `SEQSYMBOL_TABLE_SIZE` into `zstd_internal.h`. It does feel a little weird to me to have these in `zstd_internal.h`, but it does fit the note at the top of `zstd_internal.h`:

```
/* this module contains definitions which must be identical
 * across compression, decompression and dictBuilder.
 * It also contains a few functions useful to at least 2 of them
 * and which benefit from being inlined */
```

Reply:
Something I'm wondering :
a dictionary can be loaded from `lib/compress`, since it can be used for compression.
Hence `lib/compress` must be able to parse the dictionary header.

Maybe there is a way to extract the information you are looking for (header size) while using only functions already present in `lib/common` and `lib/compress` ?

Reply:
Ahh, I was wondering this when first approaching this issue, since I assumed that decompress and compress would have very similar dict parsing functions, but this is not really the case: `lib/compress` uses `ZSTD_loadZstdDictionary()` to parse the dictionary into the context and returns a dictID, while `lib/decompress` uses `ZSTD_loadDEntropy()` to parse the dictionary and returns the header size. 

`ZSTD_loadZstdDictionary()` only ephemerally computes a header size during the parsing, and it's function signature is much more complicated than `ZSTD_loadDEntropy()`:
```
static size_t ZSTD_loadZstdDictionary(ZSTD_compressedBlockState_t* bs,
                                      ZSTD_matchState_t* ms,
                                      ZSTD_cwksp* ws,
                                      ZSTD_CCtx_params const* params,
                                      const void* dict, size_t dictSize,
                                      ZSTD_dictTableLoadMethod_e dtlm,
                                      void* workspace)
```
So I think there isn't really much that's directly usable in `lib/compress`. While I think it would be theoretically possible to generate the header size from functions in `lib/common`, it would be akin to taking approach #2 I outlined in my first comment, which I don't think is the best, since `ZSTD_loadDEntropy()` already does close to the minimum amount of work required to get the dictionary header size.

Reply:
Went ahead and updated this with the first option you suggested, moving some symbols into `zstd_internal.h`. Let me know how that looks to you!